File: packages/shared/src/types/activity.ts
import type { TRACE_STEP_STATUSES } from "./enums.js";
import type { Environment, PermissionMode, RiskLevel } from "./risk.js";
import type { ToolExecutionStatus } from "./tool.js";

/**
 * Activity + execution trace.
 *
 * Activity = audit stream (who/what/where/when/why/result).
 * Trace = the Agent's step-by-step working, rendered as cards in the AI panel.
 */

export const ACTIVITY_TYPES = [
  "connection",
  "authentication",
  "configuration",
  "deployment",
  "service",
  "container",
  "file_change",
  "agent_action",
  "approval",
  "health",
] as const;
export type ActivityType = (typeof ACTIVITY_TYPES)[number];

export const ACTIVITY_SOURCES = ["agent", "user", "system", "docker", "git", "cloud"] as const;
export type ActivitySource = (typeof ACTIVITY_SOURCES)[number];

export const ACTIVITY_OUTCOMES = ["success", "failure", "cancelled", "denied"] as const;
export type ActivityOutcome = (typeof ACTIVITY_OUTCOMES)[number];

export interface Activity {
  id: string;
  serverId?: string;
  workspaceId?: string;
  type: ActivityType;
  title: string;
  description?: string;
  source: ActivitySource;
  /** Actor: which agent / which user / which subsystem ("who"). */
  actor: string;
  /** Why it happened — the agent's stated intent, or the triggering event. */
  reason?: string;
  outcome?: ActivityOutcome;
  /** Link to the trace record for drill-down. */
  traceId?: string;
  createdAt: string;
}

export interface ActivityListInput {
  serverId?: string;
  limit?: number;
}

export interface ActivityListResponse {
  activities: Activity[];
}

export type TraceStepStatus = (typeof TRACE_STEP_STATUSES)[number];

/** One visible line in "Agent working...". */
export interface TraceStep {
  traceId: string;
  stepId: string;
  seq: number;
  /** Short verb phrase for the card title, e.g. "Read application logs". */
  title: string;
  status: TraceStepStatus;
  kind: "tool" | "reasoning" | "approval" | "verification" | "note";
  toolName?: string;
  /** Raw provider-facing name is never stored; internal names only (ADR 0004). */
  input?: unknown;
  outputSummary?: string;
  outputPreview?: string;
  outputTruncated?: boolean;
  error?: string;
  startedAt?: string;
  endedAt?: string;
  durationMs?: number;
}

/** Persisted into `tool_executions` (). */
export interface ToolExecutionRecord {
  traceId: string;
  stepId: string;
  callId: string;
  toolName: string;
  serverId?: string;
  environment: Environment;
  riskLevel: RiskLevel;
  /** Which layer decided: "policy" means it was auto-approved by rules, "user" means approved in UI. */
  decision: PermissionMode;
  approvedBy?: "user" | "policy";
  status: ToolExecutionStatus;
  input: unknown;
  output?: unknown;
  error?: string;
  startedAt: string;
  endedAt?: string;
  durationMs?: number;
}

File: crates/database/src/models/activity.rs
//! activity rows and their typed columns.

use serde::{Deserialize, Serialize};

enum_as_str!(ActivityType, Connection => "connection", Authentication => "authentication", Configuration => "configuration", Deployment => "deployment", Service => "service", Container => "container", FileChange => "file_change", AgentAction => "agent_action", Approval => "approval", Health => "health");
enum_as_str!(ActivitySource, Agent => "agent", User => "user", System => "system", Docker => "docker", Git => "git", Cloud => "cloud");
enum_as_str!(ActivityOutcome, Success => "success", Failure => "failure", Cancelled => "cancelled", Denied => "denied");

/// Activity type, as sent over IPC.
///
/// `snake_case`, not `lowercase`: `FileChange` and `AgentAction` are multi-word,
/// and `lowercase` is a plain `to_ascii_lowercase()` that turned them into
/// `"filechange"`/`"agentaction"`. The shared contract
/// (`packages/shared/src/types/activity.ts`) and this file's own `as_str` /
/// `from_db` both use `"file_change"`/`"agent_action"`, so the IPC payload was
/// the only place with the wrong spelling — and nothing consumed it, which is
/// why the mismatch survived.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityType {
    Connection,
    Authentication,
    Configuration,
    Deployment,
    Service,
    Container,
    FileChange,
    AgentAction,
    Approval,
    Health,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ActivitySource {
    Agent,
    User,
    System,
    Docker,
    Git,
    Cloud,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ActivityOutcome {
    Success,
    Failure,
    Cancelled,
    Denied,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Activity {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub server_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub workspace_id: Option<String>,
    pub r#type: ActivityType,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub source: ActivitySource,
    pub actor: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub outcome: Option<ActivityOutcome>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trace_id: Option<String>,
    pub created_at: String,
}
